qdr/_ranker.cc

```diff
@@ -0,0 +1,416 @@
+        if (got != doc_counts.end())
+        {
+            double idf = get_idf(it_query->first);
+            double tf_doc = (double) got->second;
+            score += ((double) it_query->second) * idf * (BM25_K1 + 1.0) *
+                tf_doc /
+                (BM25_K1 * ((1.0 - BM25_B) + BM25_B * Ld_Lave) + tf_doc);
+        }
+        // else query word not in document
+        // in this case numerator == 0 so this contribution to score is 0
+    }
+
+    return score;
+}
+
+
+// magic constants for the language model function
+// see Zhai and Lafferty "A Study of Smoothing
+// Methods for Language Models Applied to Ad Hoc Information Retrieval"
+// in this paper, we are interested in the parameter values
+// for "title queries" (short queries).  From the conclusion,
+// 3rd to last paragraph:
+// Jelinek-Mercer
+#define LAM 0.1
+// Dirichlet
+#define MU 2000.0
+// absolute discount
+#define DELTA 0.7
+
+lm_scores_t QDR::lm(const word_counts_t& doc_counts,
+    const word_counts_t& query_counts, const double& sum_count_w_given_doc)
+{
+    /// Language model relevance
+    //    Returns:
+    //      jm: Jelinek-Mercer score,
+    //      dirichlet: Dirichlet score,
+    //      ad: Absolute Discount score
+
+    // Uses the formula in Zhai and Lafferty "A Study of Smoothing
+    // Methods for Language Models Applied to Ad Hoc Information Retrieval",
+    // SIGIR 2001.  See Table 1.
+
+    // NOTE: the original paper used PorterStemmer first and no
+    // stop words
+
+    //relevance = p(w | d) = ps(w | d) if w in document
+    //          alpha_d * p(w | C)     if w not in document
+    //
+    // ps(w | d) given in table 1
+    //  alpha_d given in table 1
+    //  p(w | C) == corpus unigram probability
+
+    // sum_w c(w; d) = total words in document = |d| (for absolute discount)
+
+    // |d|_u for absolute discount
+    double unique_terms = (double) doc_counts.size();
+
+    lm_scores_t alpha_d = {
+        LAM,
+        MU / (sum_count_w_given_doc + MU),
+        DELTA * unique_terms / sum_count_w_given_doc
+    };
+
+    // we'll return log(probability)
+    lm_scores_t ret = {0.0, 0.0, 0.0};
+    for (word_counts_t::const_iterator it_query = query_counts.begin();
+        it_query != query_counts.end(); ++it_query)
+    {
+        // need to compute probability of word given unigram language model
+        // for corpus.  using add one smoothing,
+        // P(word) = (count + 1) / (N words corpus + N words vocab + 1)
+        // (denominator: without smoothing is N words corpus,
+        //      but the +1 smoothing adds in the N words vocab + "unknown" word
+        counts_t::const_iterator got_corpus = counts.find(it_query->first);
+        uint64_t word_count_corpus = 0;
+        if (got_corpus != counts.end())
+            word_count_corpus = got_corpus->second.first;
+        double p_w_given_C = ((double) word_count_corpus + 1.0) / n2p1;
```

Comment:
I would use constructor syntax here instead of a C-style cast; however, the ` + 1.0` does do the implicit conversion for you.